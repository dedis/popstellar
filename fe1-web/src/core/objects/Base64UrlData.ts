import base64url from 'base64url';

export type Base64UrlDataState = string;

export class Base64UrlData extends String {
  public constructor(value: string) {
    super(value);

    // raise an exception if it's not base64 data
    base64url.decode(value);
  }

  public static encode(text: string | Buffer, encoding: string = 'utf8'): Base64UrlData {
    return new Base64UrlData(this.addPadding(base64url.encode(text, encoding)));
  }

  public decode(): string {
    return base64url.decode(this.valueOf());
  }

  public equals(o: Base64UrlData): boolean {
    return this.toString() === o.toString();
  }

  public toBuffer(): Buffer {
    return base64url.toBuffer(this.valueOf());
  }

  /**
   * Create Base64Url data from Base64 data
   * @param b64 the base64-encoded data
   */
  public static fromBase64(b64: string): Base64UrlData {
    return new Base64UrlData(this.addPadding(base64url.fromBase64(b64)));
  }

  public static fromBuffer(buf: Buffer): Base64UrlData {
    return new Base64UrlData(this.addPadding(base64url.encode(buf)));
  }

  private static addPadding(str: string): string {
    let paddedStr = str;
    while (paddedStr.length % 4 !== 0) {
      paddedStr += '=';
    }
    return paddedStr;
  }

  /**
   * Returns the primitive value used for representing the Base64UrlData,
   * a string
   * If you want to serialize an instance use .toState() instead
   */
  public valueOf(): string {
    return super.valueOf();
  }

  /**
   * Returns the serialized version of the base64url that can for instance be stored
   * in redux stores
   */
  public toState(): Base64UrlDataState {
    return super.valueOf();
  }

  /**
   * Deserializes a previously serializes instance of Base64Url
   */
  public static fromState(base64UrlDataState: Base64UrlDataState): Base64UrlData {
    return new Base64UrlData(base64UrlDataState);
  }
}
