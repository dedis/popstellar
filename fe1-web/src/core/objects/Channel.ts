import { Hash } from './Hash';
import { PublicKey } from './PublicKey';

export type Channel = string;
export const ROOT_CHANNEL: Channel = '/root';

export function getLaoIdFromChannel(ch: Channel): Hash {
  if (!ch || !ch.startsWith(ROOT_CHANNEL)) {
    throw new Error(`Invalid channel path: '${ch}'`);
  }

  const pathComponents = ch.split('/');
  if (pathComponents.length < 3) {
    throw new Error(`LAO is absent from channel path '${ch}'`);
  }

  return new Hash(pathComponents[2]);
}

export function channelFromIds(...args: Hash[]): Channel {
  if (args.length === 0) {
    return ROOT_CHANNEL;
  }

  return `${ROOT_CHANNEL}/${args.map((c) => c.valueOf()).join('/')}`;
}

/**
 * Returns the social channel of the given user.
 *
 * @param laoIdHash - The hash containing the laoID of the currently opened LAO
 * @param userPublicKey - The public key of the user
 */
export function getUserSocialChannel(laoIdHash: Hash, userPublicKey: PublicKey): Channel {
  return `${ROOT_CHANNEL}/${laoIdHash.valueOf()}/social/${userPublicKey}`;
}

/**
 * Returns the coin channel
 *
 * @param laoIdHash - The hash containing the laoID of the currently opened LAO
 */
export function getCoinChannel(laoIdHash: Hash): Channel {
  return `${ROOT_CHANNEL}/${laoIdHash.valueOf()}/coin`;
}

/**
 * Returns the general channel of chirps.
 *
 * @param laoIdHash - The hash containing the laoID of the currently opened LAO
 */
export function getGeneralChirpsChannel(laoIdHash: Hash): Channel {
  return `${ROOT_CHANNEL}/${laoIdHash.valueOf()}/social/chirps`;
}

/**
 * Returns the channel of all the reactions.
 *
 * @param laoIdHash - The hash containing the laoID of the currently opened LAO
 */
export function getReactionChannel(laoIdHash: Hash): Channel {
  return `${ROOT_CHANNEL}/${laoIdHash.valueOf()}/social/reactions`;
}

/**
 * Returns the channel popcha channel for authentication of specific LAO
 * @param laoId - The id of the Lao in which to authenticate
 */
export function getPopchaAuthenticationChannel(laoId: Hash): Channel {
  return `${ROOT_CHANNEL}/${laoId.valueOf()}/authentication`;
}

/** Returns the last part of the channel which is usually an event id
 * Example:
 * Input: /root/laoID/electionID
 * Output: electionID
 *
 * @param channel - The channel whose last component we want to obtain
 *
 * @remarks
 * Using this function is equivalent to making a lot of assumptions about the channel.
 * This can be brittle and using a more validation-heavy logic may be preferable.
 */
export function getLastPartOfChannel(channel: Channel): Hash {
  const channels = channel.split('/');
  return new Hash(channels[channels.length - 1]);
}
