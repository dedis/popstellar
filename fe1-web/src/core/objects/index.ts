export * from './Base64UrlData';
export * from './Channel';
export * from './Hash';
export * from './KeyPair';
export * from './PopToken';
export * from './PrivateKey';
export * from './ProtocolError';
export * from './PublicKey';
export * from './RollCallToken';
export * from './Signature';
export * from './Timestamp';
export * from './WitnessSignature';
