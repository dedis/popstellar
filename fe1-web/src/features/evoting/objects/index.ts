export * from './Election';
export * from './EncryptedVote';
export * from './Question';
export * from './QuestionResult';
export * from './RegisteredVote';
export * from './Vote';
