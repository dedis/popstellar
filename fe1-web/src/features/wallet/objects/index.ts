export * from './Seed';
export { generateToken, getCurrentPopTokenFromStore } from './Token';
export * from './Wallet';
