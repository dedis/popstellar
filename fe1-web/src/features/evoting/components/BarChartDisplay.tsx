import PropTypes, { shape } from 'prop-types';
import React from 'react';
import { StyleSheet, View, ViewStyle } from 'react-native';
import { BarChart } from 'react-native-chart-kit';

import { MajorityResult } from '../objects';

const styles = StyleSheet.create({
  barChartView: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'flex-start',
    padding: 10,
  } as ViewStyle,
});

/**
 * Displays a simple bar chart
 */
const BarChartDisplay = (props: IPropTypes) => {
  const { data } = props;

  const data2 = {
    labels: data.map((d) => d.ballotOption),
    datasets: [
      {
        data: data.map((d) => d.count),
      },
    ],
  };

  const chartConfig = {
    backgroundGradientFrom: '#1E2923',
    backgroundGradientFromOpacity: 0,
    backgroundGradientTo: '#08130D',
    backgroundGradientToOpacity: 0.3,
    height: 5000,
    color: (opacity = 1) => `rgba(0, 0, 0, ${opacity})`,
    strokeWidth: 2, // optional, default 3
    decimalPlaces: 0,
    useShadowColorFromDataset: false, // optional
  };

  // Displays a Bar Chart
  return (
    <View style={styles.barChartView}>
      <BarChart
        data={data2}
        width={460}
        height={260}
        chartConfig={chartConfig}
        yAxisLabel=""
        yAxisSuffix=""
      />
    </View>
  );
};

const propTypes = {
  data: PropTypes.arrayOf(
    shape({
      ballotOption: PropTypes.string.isRequired,
      count: PropTypes.number.isRequired,
    }),
  ).isRequired,
};
BarChartDisplay.propTypes = propTypes;

type IPropTypes = {
  data: MajorityResult[];
};

export default BarChartDisplay;
