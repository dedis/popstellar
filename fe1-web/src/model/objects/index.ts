import * as w from './wallet';

export const Wallet = w;

export * from './Base64Url';
export * from './Channel';
export * from './ConnectToLao';
export * from './Chirp';
export * from './Election';
export * from './Hash';
export * from './KeyPair';
export * from './Lao';
export * from './LaoEvent';
export * from './LaoEventBuilder';
export * from './Meeting';
export * from './PopToken';
export * from './PrivateKey';
export * from './PublicKey';
export * from './Reaction';
export * from './RollCall';
export * from './Signature';
export * from './Timestamp';
export * from './WitnessSignature';
