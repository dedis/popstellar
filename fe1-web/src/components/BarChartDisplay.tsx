import React from 'react';
import PropTypes, { shape } from 'prop-types';
import { View } from 'react-native';
import containerStyles from 'styles/stylesheets/containerStyles';
import { BarChart } from 'react-native-chart-kit';
import { MajorityResult } from 'model/objects';

/**
 * Displays a simple bar chart
 */

const BarChartDisplay = (props: IPropTypes) => {
  const { data } = props;

  const data2 = {
    labels: data.map((d) => d.ballot_option),
    datasets: [
      {
        data: data.map((d) => d.count),
      },
    ],
  };

  const chartConfig = {
    backgroundGradientFrom: '#1E2923',
    backgroundGradientFromOpacity: 0,
    backgroundGradientTo: '#08130D',
    backgroundGradientToOpacity: 0.3,
    height: 5000,
    color: (opacity = 1) => `rgba(0, 0, 0, ${opacity})`,
    strokeWidth: 2, // optional, default 3
    decimalPlaces: 0,
    useShadowColorFromDataset: false, // optional
  };

  // Displays a Bar Chart
  return (
    <View style={[containerStyles.anchoredCenter, { padding: 10, justifyContent: 'flex-start' }]}>
      <BarChart data={data2} width={460} height={260} chartConfig={chartConfig} />
    </View>
  );
};

const propTypes = {
  data: PropTypes.arrayOf(
    shape({
      ballot_option: PropTypes.string.isRequired,
      count: PropTypes.number.isRequired,
    }),
  ).isRequired,
};
BarChartDisplay.propTypes = propTypes;

type IPropTypes = {
  data: MajorityResult[];
};

export default BarChartDisplay;
