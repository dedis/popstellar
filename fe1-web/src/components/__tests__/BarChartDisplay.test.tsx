import React from 'react';
import { render } from '@testing-library/react-native';
import { MajorityResult } from 'model/objects';
import BarChartDisplay from '../BarChartDisplay';

describe('BarChartDisplay', () => {
  it('renders correctly', () => {
    const data: MajorityResult[] = [
      { ballot_option: 'Yes', count: 10 },
      { ballot_option: 'No', count: 5 },
      { ballot_option: "I don't know", count: 4 },
    ];
    const component = render(<BarChartDisplay data={data} />).toJSON();
    expect(component).toMatchSnapshot();
  });
});
