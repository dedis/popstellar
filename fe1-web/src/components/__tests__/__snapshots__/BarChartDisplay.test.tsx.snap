// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BarChartDisplay renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "space-evenly",
      },
      Object {
        "justifyContent": "flex-start",
        "padding": 10,
      },
    ]
  }
>
  <View
    style={Object {}}
  >
    <RNSVGSvgView
      bbHeight={260}
      bbWidth={460}
      focusable={false}
      height={260}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {
            "flex": 0,
            "height": 260,
            "width": 460,
          },
        ]
      }
      width={460}
    >
      <RNSVGGroup>
        <RNSVGDefs>
          <RNSVGLinearGradient
            gradient={
              Array [
                0,
                1976611,
                1,
                1292374797,
              ]
            }
            gradientTransform={null}
            gradientUnits={1}
            name="backgroundGradient"
            x1={0}
            x2={460}
            y1={5000}
            y2={0}
          />
          <RNSVGLinearGradient
            gradient={
              Array [
                0,
                436207616,
                1,
                0,
              ]
            }
            gradientTransform={null}
            gradientUnits={1}
            name="fillShadowGradient"
            x1={0}
            x2={0}
            y1={0}
            y2={5000}
          />
        </RNSVGDefs>
        <RNSVGDefs />
        <RNSVGRect
          fill={
            Array [
              1,
              "backgroundGradient",
            ]
          }
          height={260}
          propList={
            Array [
              "fill",
            ]
          }
          rx={0}
          ry={0}
          width="100%"
          x={0}
          y={0}
        />
        <RNSVGGroup>
          <RNSVGLine
            propList={
              Array [
                "stroke",
                "strokeWidth",
                "strokeDasharray",
              ]
            }
            stroke={855638016}
            strokeDasharray={
              Array [
                "5",
                "10",
              ]
            }
            strokeWidth={1}
            x1={0}
            x2={460}
            y1={16}
            y2={16}
          />
          <RNSVGLine
            propList={
              Array [
                "stroke",
                "strokeWidth",
                "strokeDasharray",
              ]
            }
            stroke={855638016}
            strokeDasharray={
              Array [
                "5",
                "10",
              ]
            }
            strokeWidth={1}
            x1={0}
            x2={460}
            y1={64.75}
            y2={64.75}
          />
          <RNSVGLine
            propList={
              Array [
                "stroke",
                "strokeWidth",
                "strokeDasharray",
              ]
            }
            stroke={855638016}
            strokeDasharray={
              Array [
                "5",
                "10",
              ]
            }
            strokeWidth={1}
            x1={0}
            x2={460}
            y1={113.5}
            y2={113.5}
          />
          <RNSVGLine
            propList={
              Array [
                "stroke",
                "strokeWidth",
                "strokeDasharray",
              ]
            }
            stroke={855638016}
            strokeDasharray={
              Array [
                "5",
                "10",
              ]
            }
            strokeWidth={1}
            x1={0}
            x2={460}
            y1={162.25}
            y2={162.25}
          />
          <RNSVGLine
            propList={
              Array [
                "stroke",
                "strokeWidth",
                "strokeDasharray",
              ]
            }
            stroke={855638016}
            strokeDasharray={
              Array [
                "5",
                "10",
              ]
            }
            strokeWidth={1}
            x1={0}
            x2={460}
            y1={211}
            y2={211}
          />
        </RNSVGGroup>
        <RNSVGGroup>
          <RNSVGText
            content={null}
            dx={Array []}
            dy={Array []}
            fill={3422552064}
            font={
              Object {
                "fontSize": 12,
                "textAnchor": "end",
              }
            }
            matrix={
              Array [
                1,
                0,
                -0,
                1,
                0,
                0,
              ]
            }
            propList={
              Array [
                "fill",
              ]
            }
            rotate={Array []}
            x={
              Array [
                52,
              ]
            }
            y={
              Array [
                211,
              ]
            }
          >
            <RNSVGTSpan
              content="4"
              dx={Array []}
              dy={Array []}
              font={Object {}}
              rotate={Array []}
              x={Array []}
              y={Array []}
            />
          </RNSVGText>
          <RNSVGText
            content={null}
            dx={Array []}
            dy={Array []}
            fill={3422552064}
            font={
              Object {
                "fontSize": 12,
                "textAnchor": "end",
              }
            }
            matrix={
              Array [
                1,
                0,
                -0,
                1,
                0,
                0,
              ]
            }
            propList={
              Array [
                "fill",
              ]
            }
            rotate={Array []}
            x={
              Array [
                52,
              ]
            }
            y={
              Array [
                162.25,
              ]
            }
          >
            <RNSVGTSpan
              content="6"
              dx={Array []}
              dy={Array []}
              font={Object {}}
              rotate={Array []}
              x={Array []}
              y={Array []}
            />
          </RNSVGText>
          <RNSVGText
            content={null}
            dx={Array []}
            dy={Array []}
            fill={3422552064}
            font={
              Object {
                "fontSize": 12,
                "textAnchor": "end",
              }
            }
            matrix={
              Array [
                1,
                0,
                -0,
                1,
                0,
                0,
              ]
            }
            propList={
              Array [
                "fill",
              ]
            }
            rotate={Array []}
            x={
              Array [
                52,
              ]
            }
            y={
              Array [
                113.5,
              ]
            }
          >
            <RNSVGTSpan
              content="7"
              dx={Array []}
              dy={Array []}
              font={Object {}}
              rotate={Array []}
              x={Array []}
              y={Array []}
            />
          </RNSVGText>
          <RNSVGText
            content={null}
            dx={Array []}
            dy={Array []}
            fill={3422552064}
            font={
              Object {
                "fontSize": 12,
                "textAnchor": "end",
              }
            }
            matrix={
              Array [
                1,
                0,
                -0,
                1,
                0,
                0,
              ]
            }
            propList={
              Array [
                "fill",
              ]
            }
            rotate={Array []}
            x={
              Array [
                52,
              ]
            }
            y={
              Array [
                64.75,
              ]
            }
          >
            <RNSVGTSpan
              content="9"
              dx={Array []}
              dy={Array []}
              font={Object {}}
              rotate={Array []}
              x={Array []}
              y={Array []}
            />
          </RNSVGText>
          <RNSVGText
            content={null}
            dx={Array []}
            dy={Array []}
            fill={3422552064}
            font={
              Object {
                "fontSize": 12,
                "textAnchor": "end",
              }
            }
            matrix={
              Array [
                1,
                0,
                -0,
                1,
                0,
                0,
              ]
            }
            propList={
              Array [
                "fill",
              ]
            }
            rotate={Array []}
            x={
              Array [
                52,
              ]
            }
            y={
              Array [
                16,
              ]
            }
          >
            <RNSVGTSpan
              content="10"
              dx={Array []}
              dy={Array []}
              font={Object {}}
              rotate={Array []}
              x={Array []}
              y={Array []}
            />
          </RNSVGText>
        </RNSVGGroup>
        <RNSVGGroup>
          <RNSVGText
            content={null}
            dx={Array []}
            dy={Array []}
            fill={3422552064}
            font={
              Object {
                "fontSize": 12,
                "textAnchor": "middle",
              }
            }
            matrix={
              Array [
                1,
                0,
                -0,
                1,
                0,
                0,
              ]
            }
            propList={
              Array [
                "fill",
              ]
            }
            rotate={Array []}
            x={
              Array [
                96,
              ]
            }
            y={
              Array [
                235,
              ]
            }
          >
            <RNSVGTSpan
              content="Yes"
              dx={Array []}
              dy={Array []}
              font={Object {}}
              rotate={Array []}
              x={Array []}
              y={Array []}
            />
          </RNSVGText>
          <RNSVGText
            content={null}
            dx={Array []}
            dy={Array []}
            fill={3422552064}
            font={
              Object {
                "fontSize": 12,
                "textAnchor": "middle",
              }
            }
            matrix={
              Array [
                1,
                0,
                -0,
                1,
                0,
                0,
              ]
            }
            propList={
              Array [
                "fill",
              ]
            }
            rotate={Array []}
            x={
              Array [
                228,
              ]
            }
            y={
              Array [
                235,
              ]
            }
          >
            <RNSVGTSpan
              content="No"
              dx={Array []}
              dy={Array []}
              font={Object {}}
              rotate={Array []}
              x={Array []}
              y={Array []}
            />
          </RNSVGText>
          <RNSVGText
            content={null}
            dx={Array []}
            dy={Array []}
            fill={3422552064}
            font={
              Object {
                "fontSize": 12,
                "textAnchor": "middle",
              }
            }
            matrix={
              Array [
                1,
                0,
                -0,
                1,
                0,
                0,
              ]
            }
            propList={
              Array [
                "fill",
              ]
            }
            rotate={Array []}
            x={
              Array [
                360,
              ]
            }
            y={
              Array [
                235,
              ]
            }
          >
            <RNSVGTSpan
              content="I don't know"
              dx={Array []}
              dy={Array []}
              font={Object {}}
              rotate={Array []}
              x={Array []}
              y={Array []}
            />
          </RNSVGText>
        </RNSVGGroup>
        <RNSVGGroup>
          <RNSVGRect
            fill={
              Array [
                1,
                "fillShadowGradient",
              ]
            }
            height={195}
            propList={
              Array [
                "fill",
              ]
            }
            rx={0}
            width={32}
            x={80}
            y={16}
          />
          <RNSVGRect
            fill={
              Array [
                1,
                "fillShadowGradient",
              ]
            }
            height={32.5}
            propList={
              Array [
                "fill",
              ]
            }
            rx={0}
            width={32}
            x={212}
            y={178.5}
          />
          <RNSVGRect
            fill={
              Array [
                1,
                "fillShadowGradient",
              ]
            }
            height={0}
            propList={
              Array [
                "fill",
              ]
            }
            rx={0}
            width={32}
            x={344}
            y={211}
          />
        </RNSVGGroup>
        <RNSVGGroup />
        <RNSVGGroup>
          <RNSVGRect
            fill={2566914048}
            height={2}
            propList={
              Array [
                "fill",
              ]
            }
            width={32}
            x={80}
            y={16}
          />
          <RNSVGRect
            fill={2566914048}
            height={2}
            propList={
              Array [
                "fill",
              ]
            }
            width={32}
            x={212}
            y={178.5}
          />
          <RNSVGRect
            fill={2566914048}
            height={2}
            propList={
              Array [
                "fill",
              ]
            }
            width={32}
            x={344}
            y={211}
          />
        </RNSVGGroup>
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
</View>
`;
